### ISSUE TITLE
Putting Graph layouting together

### ISSUE THREAD

**Issue by SimonDanisch** [abhijithanilkumar/NetworkLayout.jl#4]
Hi,
I want to put the Graph layout situation a bit in order!
We have different packages, which implement different layouting algorithms.
E.g. [GraphLayout](https://github.com/IainNZ/GraphLayout.jl), [GraphPlots](https://github.com/afternone/GraphPlot.jl/blob/master/src/layout.jl), [PlotRecipes](https://github.com/JuliaPlots/PlotRecipes.jl/blob/master/src/graphs.jl)

What I'd like to do in order to clean up the situation is to move e.g. this package to JuliaGraphs and then move the different algorithms there while preserving authorship.

While at it, I'd like to make the interface more uniform.

I'm proposing 3 things:

1.) All layouting functions should follow the same naming conventions. E.g. `Algorithm.layout(...)`. Layout should accept sparse, dense adjacency matrices and linked list of neighbors. On top of the algorithm specific keyword args, it should also accept a list of boundingboxes to take the size of nodes into account. There should also be an inplace version of layout!

2.) When possible, algorithms should possibly scale to multiple dimensions with ease. For most algorithms this is easily achieved by using `Vector{Point{N}}`, instead of `NxVector{Float}`. Note that N>3 can make sense, when visualizing the result e.g. with color as an additional dimension
To save on conversions and probably get some speed up from the `SoA` form, we can use [StructsOfArrays.jl](https://github.com/simonster/StructsOfArrays.jl), which can turn `NxVector{Float}` into `AbstractVector{Point{N}}` without any overhead.

3.) I'd like to apply the iterator pattern to iterative algorithms, so that it's easier for the user to e.g. visualize different steps of the algorithm.

In code, this might look something like that:

```Julia
module Algorithm

function layout(graph::Union{...}; algorithm_kw_args, boundingboxes, startpositions, iterations=...)
    layout!(graph, startpositions; algorithm_kw_args=algorithm_kw_args, boundingboxes=boundingboxes) 
end

# for iterative algorithms
function layout!(...; iterations= 0) # 0, or a number found to be reasonable? I'm for the latter
iterator = GraphLayouter(...)
    for i=1:iterations
        next(iterator)
    end
    positions(iterator)
end

start(::GraphLayouter) = ...
done(::GraphLayouter, state) = ...
next(::GraphLayouter, state) = # implement inner loop of iteration
positions(::GraphLayouter) = ...
# maybe also:
push!(::GraphLayouter, node) = ...
delete!(::GraphLayouter, idx) = ...

end

#usage:

positions = layout(...) # simple form
#interactive form:
iter = GraphLayouter(...) # should have the same signature as layout
while not_stopped
   next(iter)
   visualize(positions(iter))
end

```

Hope this sounds like a reasonable plan for everyone involved!

CC:
@viralbshah
@sbromberger
@jpfairbanks
@tbreloff
@IainNZ
@abhijithanilkumar 
@CarloLucibello
@afternone

**Comment by tbreloff** [comment]
Generally speaking this sounds like a great plan Simon. Happy to clean up
and consolidate what we have. Thanks for taking the lead.

**Comment by CarloLucibello** [comment]
I really like the initiative, it is useless to waste resources in so many packages. Also a big :+1: for the iterable interface 

two minor considerations: the package should not depend on any graph package,  and it should interpret matrix and sparse matrix input as an adjacency matrix, and a vector of vectors as an adjacency list.

**Comment by sbromberger** [comment]
Let me know when (and what) you'd like to move to JuliaGraphs.

**Comment by SimonDanisch** [comment]
Great! NetworkLayout should go there. Not 100% happy with the name, but I think it will do for now!

**Comment by sbromberger** [comment]
OK. Let me figure out how to do this. :)

**Comment by sbromberger** [comment]
One other thing - bikeshedding the name. If this is intended to become the canonical graph visualization tool, why not `GraphVisualization.jl` or something like that?

**Comment by tbreloff** [comment]
>  If this is intended to become the canonical graph visualization tool

I don't think this is the "visualization tool"... this is the "home for layout algorithms".  There are a million ways to draw a graph to the screen... right now we only care about consolidating those core algorithms that can give us coordinates for the placement of each node.

**Comment by sbromberger** [comment]
Ah, ok, no worries.

@SimonDanisch - I've given you admin access to the NetworkLayout.jl repo in JuliaGraphs. Do what you need to do :)

**Comment by tbreloff** [comment]
I'd like to add that I'd really prefer to avoid dependecies like JuMP for
this, but I understand that might be tough for all algorithms. Conditional
deps would be nice.

**Comment by sbromberger** [comment]
@tbreloff - is there a way to do conditional dependencies yet with Julia?

**Comment by tbreloff** [comment]
Not a great way, but I do it by just wrapping the definitions in an `is_installed` call:

```julia
function is_installed(pkgstr::AbstractString)
    try
        Pkg.installed(pkgstr) === nothing ? false: true
    catch
        false
    end
end

if is_installed("JuMP")
    @eval begin
        import JuMP
        ...
    end
end
```

**Comment by sbromberger** [comment]
Ah, IIRC that doesn't work if there are macros in the package or within the block.

**Comment by afternone** [comment]
That's cool to put them together and provide a unified interface. :+1:
